Bottom: 24f478e5ba0f7a721bb19ba9f950918132f7ada2
Top:    9323509b03f758d47a6871a871357391763d5c85
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-06 23:17:41 +0000

Fixup draw.c not to special case based on HID name, use flags instead


---

diff --git a/src/draw.c b/src/draw.c
index f9fc2c7..16d209b 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -925,12 +925,11 @@ DrawRats (const BoxType *drawn_area)
    * XXX gtk only allows negative drawing.
    * XXX using the mask here is to get rat transparency
    */
-  int can_mask = strcmp(gui->name, "lesstif") == 0;
 
-  if (can_mask)
+  if (hid_draw_can_draw_in_mask_clear (gui->graphics))
     hid_draw_use_mask (gui->graphics, HID_MASK_CLEAR);
   r_search (PCB->Data->rat_tree, drawn_area, NULL, rat_callback, NULL);
-  if (can_mask)
+  if (hid_draw_can_draw_in_mask_clear (gui->graphics))
     hid_draw_use_mask (gui->graphics, HID_MASK_OFF);
 }
 
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index 28d2fdf..1b96378 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -4137,6 +4137,8 @@ hid_lesstif_init ()
 
   lesstif_graphics_class.draw_pcb_polygon = common_gui_draw_pcb_polygon;
 
+  lesstif_graphics_class.can_draw_in_mask_clear = true;
+
   lesstif_graphics.klass = &lesstif_graphics_class;
   lesstif_graphics.poly_before = true;
   common_nogui_graphics_init (&lesstif_graphics);
diff --git a/src/hid_draw.h b/src/hid_draw.h
index a442390..7cf8c84 100644
--- a/src/hid_draw.h
+++ b/src/hid_draw.h
@@ -82,6 +82,9 @@ typedef struct hid_draw_class_st
   /* draw.c currently uses this to shortcut and special-case certain rendering when displaying on-screen */
   bool gui;
 
+  /* If set, the renderer is capable of drawing positive graphics when in HID_MASK_CLEAR mode */
+  bool can_draw_in_mask_clear;
+
 } HID_DRAW_CLASS;
 
 /* Base HID_DRAW elements visible to any module */
@@ -108,6 +111,12 @@ hid_draw_is_gui (HID_DRAW *hid_draw)
   return hid_draw->klass->gui;
 }
 
+inline bool
+hid_draw_can_draw_in_mask_clear (HID_DRAW *hid_draw)
+{
+  return hid_draw->klass->can_draw_in_mask_clear;
+}
+
 
 /* Calling wrappers to access the vfunc table */
